### PR TITLE
apply/merge: use `mark_merge_done`

### PR DESCRIPTION
**Pull request by BusyJay**
Otherwise, gc messages can be skipped when received a stale message, which will lead to orphan replica.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1745,12 +1745,12 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             let min_index = peer.get_min_progress() + 1;
             let low = cmp::max(min_index, state.get_min_index());
             // TODO: move this into raft module.
-            let entries = if low >= state.get_commit() {
+            let entries = if low >= state.get_commit() + 1 {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: PTAL @foreverbell

Comment (siddontang) [reply src/raftstore/store/store.rs]: can you explain why +1 now?

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1808,13 +1808,18 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-    fn on_ready_prepare_merge(&mut self, region: metapb::Region, state: MergeState) {
+    fn on_ready_prepare_merge(&mut self, region: metapb::Region, state: MergeState, merged: bool) {
         {
             let peer = self.region_peers.get_mut(&region.get_id()).unwrap();
             peer.pending_merge = Some(state);
             peer.mut_store().region = region.clone();
         }
 
+        if merged {
+            // PrepareMerge is executed by CommitMerge.
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: This comment confuses me, what is "PrepareMerge is executed"?

Comment (BusyJay) [reply src/raftstore/store/store.rs]: Just as its literal meaning. `CommitMerge` will try to catch up log for source region, which may execute `PrepareMerge`.

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -1221,7 +1222,7 @@ impl ApplyDelegate {
     fn load_entries_for_merge(&self, merge: &CommitMergeRequest, apply_index: u64) -> Vec<Entry> {
         // Entries from [first_index, last_index) need to be loaded.
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: Please update the comment. `... last_index]`

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -1352,10 +1345,7 @@ impl ApplyDelegate {
             region.set_start_key(source_region.get_start_key().to_vec());
         }
         write_peer_state(&self.engine, ctx.wb(), &region, PeerState::Normal)
-            .and_then(|_| {
-                // Should source_region be used?
-                write_peer_state(&self.engine, ctx.wb(), exist_region, PeerState::Tombstone)
-            })
+            .and_then(|_| mark_merge_done(&self.engine, ctx.wb(), source_region, &self.region))
```
Comment (hicqu) [inline src/raftstore/store/worker/apply.rs]: I think we need also change the strategy in `store.rs`, line 293?

File: src/raftstore/store/store.rs (hunk 1 above)
Comment (overvenus) [reply src/raftstore/store/store.rs]: Please add some comments that we now include the `PrepareMerge`.

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (siddontang) [reply src/raftstore/store/store.rs]: Suggest adding more comments here.

File: src/raftstore/store/worker/apply.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: The comment is valid. It's `[first_index, commit_index]`, which is the same as `[first_index, commit_index + 1)`.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**